### ISSUE TITLE
Generics not written on function parameters 

### ISSUE THREAD

**Issue by kigero** [HaxeFoundation/haxe#7749]
I'm writing a utility function in Haxe that adds some information to an exception before it gets thrown, and I'm having an issue with this in Java because the type parameters aren't being written out to the java function.

I've got a function defined in haxe like this :

```
public static function ex<T:java.lang.Exception>(exClass:Class<T>, msg:String):T
{
    return Type.createInstance(exClass, [msg == null ? "" : msg]);
}
```

When I comple this with haxe for the Java target, the java code that gets output looks like this:

```
public static <T> T ex(java.lang.Class exClass, java.lang.String msg)
{
    return ((T) (haxe.root.Type.createInstance(((java.lang.Class) (exClass) ), ((haxe.root.Array) (new haxe.root.Array(new java.lang.Object[]{( (( msg == null )) ? ("") : (msg) )})) ))) );
}
```
Note that the type parameters is missing from the exClass variable in the function definition.

If I then try to use this function from java:
`throw ex(IllegalArgumentException.class, "test");`
I get a compile error that "Object cannot be converted to Throwable". I believe this is because the type parameter is missing from the exClass parameter, so the java compiler just sets the return type to Object. If I add that type parameter to the exClass type:

```
public static <T> T ex2(java.lang.Class<T> exClass, java.lang.String msg)
{
	return ((T) (haxe.root.Type.createInstance(((java.lang.Class) (exClass) ), ((haxe.root.Array) (new haxe.root.Array(new java.lang.Object[]{( (( msg == null )) ? ("") : (msg) )})) ))) );
}
```

With this change the function compiles and works as expected. Any ideas if there is a workaround for this issue?

**Comment by kevinresol** [comment]
Type parameter generation for hxjava is not perfect at the moment. Maybe related: https://github.com/HaxeFoundation/haxe/issues/6289

**Comment by nadako** [comment]
@Simn I _think_ this is also a problem on JVM, because here's what the decompiler shows:

![image](https://user-images.githubusercontent.com/49749/82765227-1611f980-9e15-11ea-8557-f49bdc282225.png)

**Comment by Simn** [comment]
And what's the problem with that?

**Comment by nadako** [comment]
> If I then try to use this function from java:
> `throw ex(IllegalArgumentException.class, "test");`
> I get a compile error that "Object cannot be converted to Throwable".

Seems like an interop problem, because that `<T>` is not bound to the passed class, `javac` doesn't know if it's throwable or not.

*Disclaimer: I haven't tested it myself*

**Comment by Simn** [comment]
I remember having some issues trying to retain the type parameter of `Class<T>`, but our unit tests say it's fine to do that now. So let's try it.

Note that in this example we end up with `Class<Exception>` because we use the constraint type for singly-constrained type parameters.

**Comment by kigero** [comment]
With Haxe 4.1.1 I'm getting the same decompiled output as above, where the type parameter is `Class<Object>`, not `Class<Exception>` as you mentioned (unless I'm misunderstanding what you meant).

But I'm not sure that's the right signature in the first place.  In Haxe, I believe that if the type constraint is a class, that subclasses of the constraint are also allowed: for example 
```
public static function ex<T:java.lang.Exception>(exClass:Class<T>, msg:String):T
{
    return Type.createInstance(exClass, [msg == null ? "" : msg]);
}

public static function main():Void
{
    throw ex(java.lang.IllegalArgumentException, "test");
}
```
This works because `IllegalArgumentException `is a subclass of `Exception`.  In Java though, subclasses are not allowed - here's the function output by Haxe (although I did change the `Class<Object>` to `Class<Exception>`) and an equivalent invocation from Java:
```
public static <T> Exception ex(Class<Exception> exClass, String msg)
{
    return (Exception)Type.createInstance(exClass, Array.ofNative(new Object[] { msg == null ? "" : msg }));
}

public static void main(String[] args) throws Exception
{
    throw ex(IllegalArgumentException.class, "test");
}
```
This won't compile, giving an "incompatible types" error because `IllegalArgumentException `can't be used since  the generic type specifies `Exception`.

So even if the type parameter is inferred as `Exception`, the output Java code is still more restrictive than the Haxe code, and I'm not sure that's expected.  Am I missing something here?

**Comment by Gama11** [comment]
The commit that closed this is not in 4.1.1, it was only made yesterday.

**Comment by kigero** [comment]
Blast, missed that - I tried the nightly and that does indeed fix the first issue, I now get `Class<Exception>`.  Sorry about that.

It doesn't fix the second issue though, and I'm not sure if I should be expecting it to.  Inferring `Class<Exception>` in Java restricts callers to only the Exception class instead of subclasses as it does in Haxe.  So I'm not sure if I'm doing something wrong on the Haxe side or if there really is an issue with the generated signature.
